micrometer-test/src/main/java/io/micrometer/core/instrument/HttpClientTimingInstrumentationVerificationSuite.java

```diff
@@ -0,0 +1,131 @@
+     * {@literal http://localhost:8080}, the templatedPath would be
+     * {@literal /cart/{cartId}}. One string pathVariables argument is expected for
+     * substituting the cartId path variable. The number of pathVariables arguments SHOULD
+     * exactly match the number of path variables in the templatedPath.
+     * @param method http method to use to send the request
+     * @param baseUrl portion of the URL before the path where to send the request
+     * @param templatedPath the path portion of the URL after the baseUrl, starting with a
+     * forward slash, and optionally containing path variable placeholders
+     * @param pathVariables optional variables to substitute into the templatedPath
+     */
+    abstract void sendHttpRequest(HttpMethod method, URI baseUrl, String templatedPath, String... pathVariables);
+
+    /**
+     * Convenience method provided to substitute the template placeholders for the
+     * provided path variables. The number of pathVariables argument SHOULD match the
+     * number of placeholders in the templatedPath.
+     * @param templatedPath a URL path optionally containing placeholders in curly
+     * brackets
+     * @param pathVariables path variable values for which placeholders should be
+     * substituted
+     * @return path string with substitutions, if any, performed
+     */
+    protected String substitutePathVariables(String templatedPath, String... pathVariables) {
+        if (pathVariables.length == 0) {
+            return templatedPath;
+        }
+        String substituted = templatedPath;
+        for (String substitution : pathVariables) {
+            substituted = substituted.replaceFirst("\\{.*?}", substitution);
+        }
+        return substituted;
+    }
+
+    @Test
+    void getTemplatedPath(WireMockRuntimeInfo wmRuntimeInfo) {
+        stubFor(get(anyUrl()).willReturn(ok()));
+
+        String templatedPath = "/customers/{customerId}/carts/{cartId}";
+        sendHttpRequest(HttpMethod.GET, URI.create(wmRuntimeInfo.getHttpBaseUrl()), templatedPath, "112", "5");
+
+        Timer timer = getRegistry().get(timerName()).tags("method", "GET", "status", "200", "uri", templatedPath)
+                .timer();
+        assertThat(timer.count()).isEqualTo(1);
+        assertThat(timer.totalTime(TimeUnit.NANOSECONDS)).isPositive();
+    }
+
+    @Test
+    void unmappedUrisAreCardinalityLimited(WireMockRuntimeInfo wmRuntimeInfo) {
+        stubFor(get(anyUrl()).willReturn(notFound()));
+
+        String templatedPath = "/notFound404";
+        sendHttpRequest(HttpMethod.GET, URI.create(wmRuntimeInfo.getHttpBaseUrl()), templatedPath);
+
+        Timer timer = getRegistry().get(timerName()).tags("method", "GET", "status", "404").timer();
+        // we should standardize on a value e.g. NOT_FOUND, but for backwards
+        // compatibility, assert is more lenient
+        assertThat(timer.getId().getTag("uri")).isNotEqualTo(templatedPath);
```

Comment:
The Apache HTTP client instrumentation is failing this assertion now. We should fix that.

Reply:
While we should definitely test and enforce this for HTTP server instrumentation, it's not the same situation for HTTP client instrumentation. Maybe relaxing this is better.